client/core/core.go

```diff
@@ -4563,16 +4565,40 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 		return fmt.Errorf("match proof note unmarshal error: %w", err)
 	}
 
-	// Expire the epoch
-	dc.setEpoch(note.MarketID, note.Epoch+1)
+	// Expire previous epoch.
+	dc.tryUpdateEpoch(note.MarketID, note.Epoch+1)
 
 	book := dc.bookie(note.MarketID)
 	if book == nil {
 		return fmt.Errorf("no order book found with market id %q",
 			note.MarketID)
 	}
 
-	return book.ValidateMatchProof(note)
+	err = book.ValidateMatchProof(note)
+	if err != nil {
+		return fmt.Errorf("match proof validation failed: %w", err)
+	}
+
+	// Validate server match checksum for client orders (trades) in this epoch.
+	for _, trade := range dc.trackedTrades() {
+		// TODO - len chkeck, why ?
+		if trade.metaData.Status == order.OrderStatusEpoch && len(trade.csum) > 0 && !bytes.Equal(note.CSum, trade.csum) {
+			// Validation can fail either due to server trying to cheat
+			// (by seeing preimage before committing to csum for a match), or
+			// client loosing (e.g. due to restarting, since we don't
+			// persistently store these at the moment) his csums for the trades
+			// involved.
+			c.log.Warnf("validating match checksum for epoch %d, "+
+				"match commit checksum %s is not of the same value as "+
+				"client's order commit checksum %s, order ID: %s",
+				note.Epoch, note.CSum, trade.csum, trade.ID())
+
+			// Client doesn't want to proceed with the match for this order.
+			trade.revoke()
```

Comment:
To start I only want to warn.  We've done nothing so far, so we should at least use this in production with just a warning until we're sure we're not going to be bailing out on orders incorrectly, which will cause swaps to fail and users to be penalized (account suspended)

Reply:
Okay, warning it is for now.